File: lib/src/log.js
// Логер для сервісів без Fastify. Один рядок — один JSON: так GlitchTip і
// docker logs читають однаково, а grep по service лишається можливим.
export function makeLog(service) {
  const write = (level, msg, extra) => {
    const line = { t: new Date().toISOString(), level, service, msg, ...extra };
    process.stdout.write(`${JSON.stringify(line)}\n`);
  };
  return {
    info: (msg, extra) => write("info", msg, extra),
    warn: (msg, extra) => write("warn", msg, extra),
    error: (msg, extra) => write("error", msg, extra instanceof Error
      ? { err: extra.message, stack: extra.stack }
      : extra),
  };
}

File: backend/worker/worker/log.py
"""Лог у stdout одним рядком на подію.

Докер збирає stdout, journald його обертає, glitchtip ловить помилки —
свого файлового логу воркеру не треба. Формат людський, а не JSON: цей лог
читає людина, коли черга відстала, а не парсер.
"""
import sys
import time


def _line(level: str, msg: str, **kv: object) -> None:
    ts = time.strftime("%H:%M:%S", time.localtime())
    tail = " ".join(f"{k}={v}" for k, v in kv.items() if v is not None)
    print(f"{ts} {level} {msg}{' ' + tail if tail else ''}", file=sys.stdout, flush=True)


def info(msg: str, **kv: object) -> None:
    _line("·", msg, **kv)


def warn(msg: str, **kv: object) -> None:
    _line("!", msg, **kv)


def error(msg: str, **kv: object) -> None:
    _line("✗", msg, **kv)
